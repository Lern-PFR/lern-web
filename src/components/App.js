function App() {
  return (
    <div className="App">
      <header className="App-header">
        <p>
          Edit <code>src/components/App.js</code> and save to reload.
        </p>
      </header>
    </div>
  );
}

export default App;
