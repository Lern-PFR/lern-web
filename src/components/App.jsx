import { Home } from 'react-feather';
import { MainLayout, StyledDiv } from 'components/shared/layout';
import {
	Canon,
	Trafalgar,
	Paragon,
	GreatPrimer,
	DoublePica,
	BodyCopy,
	Pica,
	LongPrimer,
	Brevier,
	Minion,
} from 'components/shared/typography';
import {
	PrimaryButton,
	StandardButton,
	SubtleButton,
	DropdownButton,
	DangerButton,
	IconButton,
	FloatingActionButton,
} from 'components/shared/buttons';
import {
	elevationFlat,
	elevationFirst,
	elevationSecond,
	elevationThird,
	elevationFourth,
	elevationFifth,
} from 'theme/elevations';
import {
	LabeledCheckbox,
	FieldsetComponent,
	LabeledRadioButton,
	LabelComponent,
	LabeledInput,
} from 'components/shared/form';

/**
 * @name App
 * @description The main component of the application.
 *
 * @author TimothÃ©e Simon-Franza
 */
const App = () => (
	<MainLayout>
		<StyledDiv position="relative">
			<Canon>Canon text element.</Canon>
			<Trafalgar>Trafalgar text element.</Trafalgar>
			<Paragon>Paragon text element.</Paragon>
			<GreatPrimer>GreatPrimer text element.</GreatPrimer>
			<DoublePica>DoublePica text element.</DoublePica>
			<BodyCopy>BodyCopy text element.</BodyCopy>
			<Pica>Pica text element.</Pica>
			<LongPrimer>LongPrimer text element.</LongPrimer>
			<Brevier>Brevier text element.</Brevier>
			<Minion>Minion text element.</Minion>
		</StyledDiv>
		<StyledDiv position="relative">
			<PrimaryButton>Primary button</PrimaryButton>
			<StandardButton>Standard button</StandardButton>
			<SubtleButton>Subtle button</SubtleButton>
			<DropdownButton>Dropdown button</DropdownButton>
			<DangerButton>Danger button</DangerButton>
			<IconButton><Home /></IconButton>
			<FloatingActionButton isIconOnly><Home /></FloatingActionButton>
		</StyledDiv>
		<StyledDiv position="relative" marginTop="50px" display="flex" width="100%">
			<StyledDiv {...elevationFlat} width="260px" height="260px" padding="10px">Flat component</StyledDiv>
			<StyledDiv {...elevationFirst} width="260px" height="260px" padding="10px">5% elevation component</StyledDiv>
			<StyledDiv {...elevationSecond} width="260px" height="260px" padding="10px">10% elevation component</StyledDiv>
			<StyledDiv {...elevationThird} width="260px" height="260px" padding="10px">20% elevation component</StyledDiv>
			<StyledDiv {...elevationFourth} width="260px" height="260px" padding="10px">30% elevation component</StyledDiv>
			<StyledDiv {...elevationFifth} width="260px" height="260px" padding="10px">50% elevation component</StyledDiv>
		</StyledDiv>
		<StyledDiv position="relative" display="flex" flexDirection="column">
			<LabelComponent>Default label component</LabelComponent>
			<LabelComponent textStyle="bodycopy">BodyCopy label component</LabelComponent>
		</StyledDiv>
		<StyledDiv position="relative">
			<FieldsetComponent title="Basic checkbox components">
				<LabeledCheckbox id="lcb5">Labeled checkbox</LabeledCheckbox>
				<LabeledCheckbox labelTextStyle="bodycopy" checked id="lcb6">Default checked labeled checkbox</LabeledCheckbox>
				<LabeledCheckbox disabled id="lcb7">Disabled labeled checkbox</LabeledCheckbox>
				<LabeledCheckbox disabled checked id="lcb8">Disabled default checked labeled checkbox</LabeledCheckbox>
			</FieldsetComponent>
		</StyledDiv>
		<StyledDiv position="relative">
			<FieldsetComponent title="Basic radio button components">
				<LabeledRadioButton name="rg2" id="lrb5">Labeled radio button</LabeledRadioButton>
				<LabeledRadioButton labelTextStyle="bodycopy" name="rg2" checked id="lrb6">Default checked labeled radio button</LabeledRadioButton>
				<LabeledRadioButton name="rg2" disabled id="lrb7">Disabled labeled radio button</LabeledRadioButton>
			</FieldsetComponent>
		</StyledDiv>
		<StyledDiv position="relative" width="500px">
			<FieldsetComponent title="Text input components">
				<LabeledInput id="basicTextInput">Labeled text input</LabeledInput>
				<LabeledInput id="placeholderTextInput" placeholder="Enter details here.">Labeled text input with custom placeholder</LabeledInput>
				<LabeledInput id="hintTextInput" placeholder="Enter details here." hintText="This is a hint for the user.">Labeled text input with hint</LabeledInput>
				<LabeledInput
					id="disabledTextInput"
					placeholder="Enter details here."
					hintText="This is a hint for the user."
					disabled
				>
					Disabled labeled text input
				</LabeledInput>
				<LabeledInput
					id="errorTextInput"
					placeholder="Enter details here."
					hintText="This is a hint for the user."
					errorText="There is something wrong with the value in this input"
					error
				>
					Labeled text input with a validation error
				</LabeledInput>
				<LabeledInput id="numberInput" placeholder="Number Input" type="number">Labeled number input</LabeledInput>
				<LabeledInput id="passwordInput" placeholder="Password Input" type="password">Labeled password input</LabeledInput>
			</FieldsetComponent>
		</StyledDiv>
	</MainLayout>
);

export default App;
