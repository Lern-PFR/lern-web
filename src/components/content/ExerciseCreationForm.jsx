import { useCallback, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';

import { isFormValid, validateField, validateForm } from 'lib/shared/formUtils';
import { hasMaxLength, hasMinLength, isRequired } from 'lib/shared/inputValidation';

import { PrimaryButton } from 'components/shared/buttons';
import { LabeledTextArea } from 'components/shared/form';
import { StyledDiv, StyledForm } from 'components/shared/styledElements';

import { form, formSubmit, responseDiv } from 'theme/contentEditionCommon/genericLayout';
import { GreatPrimer } from 'components/shared/typography';
import ResponseItem from './ResponseItem';

/**
 * @constant
 * @name initialState
 * @description The form's initial state value.
 *
 * @author Timothée Simon-Franza
 */
const initialState = {
	title: '',
	description: '',
	content: '',
};

/**
 * @constant
 * @name validationRules
 * @description The validation rules for each field.
 *
 * @author Timothée Simon-Franza
 */
const validationRules = {
	instructions: {
		required: isRequired('required'),
		hasMinLength: hasMinLength(10, 'min_length'),
		hasMaxLength: hasMaxLength(300, 'max_length'),
	},
	explanation: {
		hasMaxLength: hasMaxLength(300, 'max_length'),
	},
};

/**
 * @constant
 * @name inputsDefinition
 * @description The form's input definition object.
 *
 * @author Christopher Walker
 */
const inputsDefinition = {
	instructions: { id: 'instructions', name: 'instructions', inputType: 'textarea', hasPlaceholder: true },
	explanation: { id: 'explanation', name: 'explanation', inputType: 'textarea', hasPlaceholder: true },
};

/**
 * @constant
 * @name inputsDefinition
 * @description The form's input definition object.
 *
 * @author Christopher Walker
 */
const responseFieldsDefinition = {
	answer1: { id: 'answer1', name: 'answer1', inputType: 'responseItem', hasPlaceholder: true, isCorrect: true },
	answer2: { id: 'answer2', name: 'answer2', inputType: 'responseItem', hasPlaceholder: true, isCorrect: false },
	answer3: { id: 'answer3', name: 'answer3', inputType: 'responseItem', hasPlaceholder: true, isCorrect: false },
};

/**
 * @name ExerciseCreationForm
 * @description A form component used to create a lesson.
 *
 * @author Christopher Walker
 *
 * @param {func} onSubmit The method to trigger on form submit.
 */
const ExerciseCreationForm = ({ onSubmit }) => {
	const { t } = useTranslation();
	const [formState, setFormState] = useState(initialState);
	const [errors, setErrors] = useState({});
	const fieldsRef = useRef({});
	const [validAnswer, setValidAnswer] = useState('answer1');

	const hasFormBeenSubmitted = useRef(false);

	/**
	 * @function
	 * @name handleSubmit
	 * @description Handles the form's submit event, performs validation and calls the onSubmit method if all is in order.
	 *
	 * @author Timothée Simon-Franza
	 *
	 * @param {object} event	The event originating from the form's submission.
	 */
	const handleSubmit = useCallback((event) => {
		event?.preventDefault();

		hasFormBeenSubmitted.current = true;
		const formValidationResult = validateForm(fieldsRef.current, validationRules);
		setErrors(formValidationResult);

		if (!isFormValid(formValidationResult)) {
			return;
		}
		const result = {
			statement: formState.instructions,
			explanation: formState.explanation,
			type: 'SingleChoice',
			answers: [
				{
					text: formState.answer1,
					valid: validAnswer === 'answer1',
				},
				{
					text: formState.answer2,
					valid: validAnswer === 'answer2',
				},
				{
					text: formState.answer3,
					valid: validAnswer === 'answer3',
				},
			],
		};
		onSubmit({ ...result });
	}, [formState, onSubmit, validAnswer]);

	/**
	 * @function
	 * @name handleChange
	 * @description Updates the formState object with the new value.
	 *
	 * @author Timothée Simon-Franza
	 *
	 * @param {object} event				The onChange event.
	 * @param {object} event.target			The target which the onChange event originated from.
	 * @param {string} event.target.name	The name of the input the event originated from.
	 * @param {string} event.target.value	The new value to update the state with.
	 */
	const handleChange = useCallback((event) => {
		event.preventDefault();
		const { target: { name, value: newValue } } = event;

		setFormState((prevState) => ({ ...prevState, [name]: newValue }));

		// Prevents the validation error message to be displayed while the user types in for the first time.
		if (hasFormBeenSubmitted.current || errors[name]) {
			const { [name]: removedError, ...rest } = errors; 								// Removes previous errors for current field.
			setErrors({ ...rest, [name]: validateField(newValue, validationRules[name]) });	// Performs validation check on the field and updates the errors state.
		}
	}, [errors, setErrors, setFormState]);

	const handleValidAnswerChange = useCallback((name) => {
		setValidAnswer(name);
	}, [setValidAnswer]);

	/**
	 * @function
	 * @name handleBlur
	 * @description Performs validation checks for the input generating the blur event.
	 *
	 * @author Timothée Simon-Franza
	 *
	 * @param {object} event				The onBlur event.
	 * @param {object} event.target			The target which the onBlur event originated from.
	 * @param {string} event.target.name	The name of the input the event originated from.
	 * @param {string} event.target.value	The current value of the input the event originated from.
	 */
	const handleBlur = useCallback((event) => {
		const { target: { name, value } } = event;

		// Removes previous errors for current field.
		const { [name]: removedError, ...rest } = errors;

		// Performs validation check on the field and updates the errors state.
		setErrors({ ...rest, [name]: validateField(value, validationRules[name]) });
	}, [errors, setErrors]);

	/**
	 *
	 * @param {*} fieldName
	 * @returns
	 */
	const getErrorMessageByFieldName = (fieldName) => {
		if (errors[fieldName] && Object.keys(errors[fieldName])?.length > 0) {
			return t(`exercises.creation.form.fields.${fieldName}.validation_rules.${Object.values(errors[fieldName])?.[0]}`);
		}

		return '';
	};

	return (
		<StyledForm onSubmit={handleSubmit} {...form}>
			{Object.values(inputsDefinition).map(({ id, name, inputType = 'text', hasPlaceholder }) => (
				<LabeledTextArea
					key={id}
					id={id}
					name={name}
					type={inputType}
					onChange={handleChange}
					onBlur={handleBlur}
					rows={6}
					hasError={errors[name] && Object.keys(errors[name])?.length > 0}
					errorText={getErrorMessageByFieldName(name)}
					placeholder={hasPlaceholder ? t(`exercises.creation.form.fields.${name}.placeholder`) : undefined}
					ref={(fieldRef) => { fieldsRef.current[name] = fieldRef; }}
					defaultValue={formState[name]}
				>
					{t(`exercises.creation.form.fields.${name}.label`)}
				</LabeledTextArea>
			))}
			<StyledDiv {...responseDiv}>
				<GreatPrimer>{t('exercises.creation.responses')}</GreatPrimer>
				{Object.values(responseFieldsDefinition).map(({ id, name, hasPlaceholder, isCorrect }) => (
					<ResponseItem
						key={id}
						id={id}
						name={name}
						hasPlaceholder={hasPlaceholder}
						handleCheckedChanged={handleValidAnswerChange}
						handleChange={handleChange}
						handleBlur={handleBlur}
						hasError={errors[name] && Object.keys(errors[name])?.length > 0}
						errorText={getErrorMessageByFieldName(name)}
						checked={isCorrect}
					/>
				))}
			</StyledDiv>
			<PrimaryButton type="submit" {...formSubmit}>
				{t('exercises.creation.form.action.submit')}
			</PrimaryButton>
		</StyledForm>
	);
};

ExerciseCreationForm.propTypes = {
	onSubmit: PropTypes.func.isRequired,
};

export default ExerciseCreationForm;
