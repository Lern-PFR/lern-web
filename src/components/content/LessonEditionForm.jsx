import { useCallback, useMemo, useRef, useState } from 'react';
import { useTranslation } from 'react-i18next';
import PropTypes from 'prop-types';

import { hasMaxLength, hasMinLength, isRequired } from 'lib/shared/inputValidation';
import { isFormValid, validateField, validateForm } from 'lib/shared/formUtils';
import { StyledForm } from 'components/shared/styledElements';
import { form, formSubmit } from 'theme/contentEditionCommon/genericLayout';
import { LabeledInput, LabeledSelect, LabeledTextArea } from 'components/shared/form';
import { PrimaryButton } from 'components/shared/buttons';

/**
 * @constant
 * @name validationRules
 * @description The validation rules for each field.
 *
 * @author Timothée Simon-Franza
 */
const validationRules = {
	title: {
		required: isRequired('required'),
		hasMinLength: hasMinLength(3, 'min_length'),
		hasMaxLength: hasMaxLength(50, 'max_length'),
	},
	description: {
		required: isRequired('required'),
		hasMinLength: hasMinLength(10, 'min_length'),
		hasMaxLength: hasMaxLength(300, 'max_length'),
	},
	content: {
		required: isRequired('required'),
	},
	order: {
		required: isRequired('required'),
	},
};

/**
 * @constant
 * @name inputsDefinition
 * @description The form's input definition object.
 *
 * @author Timothée Simon-Franza
 */
const inputsDefinition = {
	title: { id: 'title', name: 'title', hasPlaceholder: true },
	description: { id: 'description', name: 'description', inputType: 'textarea', hasPlaceholder: true },
	content: { id: 'content', name: 'content', inputType: 'textarea', hasPlaceholder: false },
	order: { id: 'order', name: 'order', inputType: 'select', hasPlaceholder: false },
};

/**
 * @name LessonEditionForm
 * @description A form component used to edit a lesson.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object}	lesson				The lesson to edit.
 * @param {array}	lessonOrderOptions	Options to provide the order select input.
 * @param {func}	onSubmit			The method to trigger on form submission.
 */
const LessonEditionForm = ({ lesson, conceptOrderOptions, onSubmit }) => {
	const initialState = useMemo(() => ({
		title: lesson.title,
		description: lesson.description,
		content: lesson.content,
		order: (lesson.order !== null && lesson.order !== undefined) ? { label: lesson.order, value: lesson.order } : null,
	}), [lesson]);

	const { t } = useTranslation();
	const [formState, setFormState] = useState(initialState);
	const [errors, setErrors] = useState({});
	const fieldsRef = useRef({});
	const hasFormBeenSubmitted = useRef(false);

	/**
	 * @function
	 * @name handleSubmit
	 * @description Handles the form's submit event, performs validation and calls the onSubmit method if all is in order.
	 *
	 * @author Timothée Simon-Franza
	 *
	 * @param {object} event The event originating from the form's submission.
	 */
	const handleSubmit = useCallback((event) => {
		event?.preventDefault();

		hasFormBeenSubmitted.current = true;
		const formValidationResult = validateForm(fieldsRef.current, validationRules);
		setErrors(formValidationResult);

		if (!isFormValid(formValidationResult)) {
			return;
		}

		const result = {
			...formState,
			order: formState.order.value ?? formState.order,
		};

		onSubmit(result);
	}, [formState, onSubmit, setErrors]);

	/**
	 * @function
	 * @name handleChange
	 * @description Updates the formState object with the new value.
	 *
	 * @author Timothée Simon-Franza
	 *
	 * @param {object} event				The onChange event.
	 * @param {object} event.target			The target which the onChange event originated from.
	 * @param {string} event.target.name	The name of the input the event originated from.
	 * @param {string} event.target.value	The new value to update the state with.
	 */
	const handleChange = useCallback((event) => {
		const { target: { name, value: newValue } } = event;

		setFormState((prevState) => ({ ...prevState, [name]: newValue }));

		// Prevents the validation error message to be displayed while the user types in for the first time.
		if (hasFormBeenSubmitted.current || errors[name]) {
			const { [name]: removedError, ...rest } = errors; 								// Removes previous errors for current field.
			setErrors({ ...rest, [name]: validateField(newValue, validationRules[name]) });	// Performs validation check on the field and updates the errors state.
		}
	}, [errors, setErrors, setFormState]);

	/**
	 * @function
	 * @name handleOrderSelectChange
	 * @description An intermediate handleChange method for the order select field.
	 *
	 * @author Timothée Simon-Franza
	 */
	const handleOrderSelectChange = useCallback(({ value }) => {
		handleChange({ target: { name: 'order', value } });
	}, [handleChange]);

	/**
	 * @function
	 * @name handleBlur
	 * @description Performs validation checks for the input generating the blur event.
	 *
	 * @author Timothée Simon-Franza
	 *
	 * @param {object} event				The onBlur event.
	 * @param {object} event.target			The target which the onBlur event originated from.
	 * @param {string} event.target.name	The name of the input the event originated from.
	 * @param {string} event.target.value	The current value of the input the event originated from.
	 */
	const handleBlur = useCallback((event) => {
		const { target: { name, value } } = event;

		// Removes previous errors for current field.
		const { [name]: removedError, ...rest } = errors;

		// Performs validation check on the field and updates the errors state.
		setErrors({ ...rest, [name]: validateField(value, validationRules[name]) });
	}, [errors, setErrors]);

	/**
	 * @function
	 * @name getErrorMessageByFieldName
	 * @description Returns the translated validation error message for the provided fieldname if it exists.
	 *
	 * @author Timothée Simon-Franza
	 *
	 * @param {string} fieldName The name of the field we want the validation message of.
	 *
	 * @returns {string}
	 */
	const getErrorMessageByFieldName = (fieldName) => {
		if (errors[fieldName] && Object.keys(errors[fieldName])?.length > 0) {
			return t(`lessons.edition.form.fields.${fieldName}.validation_rules.${Object.values(errors[fieldName])?.[0]}`);
		}

		return '';
	};

	return (
		<StyledForm onSubmit={handleSubmit} {...form}>
			{Object.values(inputsDefinition).map(({ id, name, inputType = 'text', hasPlaceholder }) => {
				switch (inputType) {
					case 'textarea': {
						return (
							<LabeledTextArea
								key={id}
								id={id}
								name={name}
								type={inputType}
								onChange={handleChange}
								rows={6}
								onBlur={handleBlur}
								hasError={errors[name] && Object.keys(errors[name])?.length > 0}
								errorText={getErrorMessageByFieldName(name)}
								placeholder={hasPlaceholder ? t(`lessons.edition.form.fields.${name}.placeholder`) : undefined}
								ref={(fieldRef) => { fieldsRef.current[name] = fieldRef; }}
								defaultValue={formState[name]}
							>
								{t(`lessons.edition.form.fields.${name}.label`)}
							</LabeledTextArea>
						);
					}
					case 'select': {
						return (
							<LabeledSelect
								key={id}
								id={id}
								name={name}
								type={inputType}
								onChange={handleOrderSelectChange}
								onBlur={handleBlur}
								options={conceptOrderOptions}
								hasError={errors[name] && Object.keys(errors[name])?.length > 0}
								errorText={getErrorMessageByFieldName(name)}
								placeholder={hasPlaceholder ? t(`lessons.edition.form.fields.${name}.placeholder`) : undefined}
								ref={(fieldRef) => { fieldsRef.current[name] = fieldRef; }}
								defaultValue={formState[name]}
								required
							>
								{t(`lessons.edition.form.fields.${name}.label`)}
							</LabeledSelect>
						);
					}
					default: {
						return (
							<LabeledInput
								key={id}
								id={id}
								name={name}
								type={inputType}
								onChange={handleChange}
								onBlur={handleBlur}
								hasError={errors[name] && Object.keys(errors[name])?.length > 0}
								errorText={getErrorMessageByFieldName(name)}
								placeholder={hasPlaceholder ? t(`lessons.edition.form.fields.${name}.placeholder`) : undefined}
								ref={(fieldRef) => { fieldsRef.current[name] = fieldRef; }}
								defaultValue={formState[name]}
							>
								{t(`lessons.edition.form.fields.${name}.label`)}
							</LabeledInput>
						);
					}
				}
			})}
			<PrimaryButton type="submit" {...formSubmit}>
				{t('lessons.edition.form.action.submit')}
			</PrimaryButton>
		</StyledForm>
	);
};

LessonEditionForm.propTypes = {
	lesson: PropTypes.shape({
		title: PropTypes.string.isRequired,
		description: PropTypes.string.isRequired,
		content: PropTypes.string.isRequired,
		order: PropTypes.number.isRequired,
	}).isRequired,
	lessonOrderOptions: PropTypes.arrayOf(
		PropTypes.shape({
			value: PropTypes.oneOfType([
				PropTypes.string,
				PropTypes.number,
			]).isRequired,
			label: PropTypes.oneOfType([
				PropTypes.string,
				PropTypes.number,
				PropTypes.node,
			]).isRequired,
		}),
	).isRequired,
	onSubmit: PropTypes.func.isRequired,
};

export default LessonEditionForm;
