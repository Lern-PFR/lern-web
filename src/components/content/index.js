import NavigationSidebar from './NavigationSidebar';
import SubjectCreationForm from './SubjectCreationForm';
import SubjectEditionForm from './SubjectEditionForm';
import SubjectModuleList from './SubjectModuleList';

export {
	NavigationSidebar,
	SubjectCreationForm,
	SubjectEditionForm,
	SubjectModuleList,
};
