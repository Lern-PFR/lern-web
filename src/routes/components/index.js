import RenderRoutes from './RenderRoutes';

export {
	// TODO: disable this exception once there is more to export
	// eslint-disable-next-line import/prefer-default-export
	RenderRoutes,
};
