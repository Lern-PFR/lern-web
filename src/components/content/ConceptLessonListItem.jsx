import { useCallback, useMemo } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { generatePath } from 'react-router';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';

import routes from 'routes';
import { history } from 'routes/components/RouterProvider';

import { StyledDiv, StyledListItem } from 'components/shared/styledElements';
import { BodyCopy, Brevier } from 'components/shared/typography';
import { OutlinedButton, OutlinedLinkButton } from 'components/shared/buttons';
import Link from 'components/shared/navigation/Link';

import { innerContentListItem, innerContentListItemActionsRow, moduleCard } from 'theme/contentEditionCommon/innerContentListTheme';
import { deleteLesson } from 'redux/actions/lessons';

const ListCard = styled(StyledListItem)(
	{
		...innerContentListItem,
		'&, & *': {
			cursor: 'pointer',
		},
	}
);

/**
 * @name ConceptLessonListItem
 * @description A component used to display a concept's lesson in the concept edition page.
 *
 * @author TimothÃ©e Simon-Franza
 *
 * @param {string} description	The current lesson's description.
 * @param {string} id			The current lesson's id.
 * @param {string} title		THe current lesson's title.
 */
const ConceptLessonListItem = ({ description, id, title }) => {
	const dispatch = useDispatch();
	const { t } = useTranslation();

	const editPageRedirectionLink = useMemo(() => generatePath(routes.lessons.lessonEdition, { lessonId: id }), [id]);

	const onCardClick = useCallback(() => {
		history.push(editPageRedirectionLink);
	}, [editPageRedirectionLink]);

	const onDeleteBtnClick = useCallback((e) => {
		if (e?.stopPropagation) {
			e.stopPropagation();
		}

		dispatch(deleteLesson(id));
	}, [dispatch, id]);

	return (
		<ListCard key={id} onClick={onCardClick} {...moduleCard}>
			<BodyCopy margin="0">{title}</BodyCopy>
			<Brevier margin="0">{description}</Brevier>
			<StyledDiv {...innerContentListItemActionsRow}>
				<OutlinedLinkButton>
					<Link to={editPageRedirectionLink}>{t('concepts.edition.lessons_list.links.edit')}</Link>
				</OutlinedLinkButton>
				<OutlinedButton onClick={onDeleteBtnClick}>
					{t('concepts.edition.lessons_list.links.delete')}
				</OutlinedButton>
			</StyledDiv>
		</ListCard>
	);
};

ConceptLessonListItem.propTypes = {
	description: PropTypes.string.isRequired,
	id: PropTypes.string.isRequired,
	title: PropTypes.string.isRequired,
};

export default ConceptLessonListItem;
