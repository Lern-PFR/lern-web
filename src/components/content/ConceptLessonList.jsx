import { useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import PropTypes from 'prop-types';
import { PlusCircle } from 'react-feather';
import { generatePath } from 'react-router';
import styled from 'styled-components';

import routes from 'routes';
import { StyledDiv, StyledList, StyledListItem } from 'components/shared/styledElements';
import { DoublePica } from 'components/shared/typography';
import { history } from 'routes/components/RouterProvider';

import { addInnerContentCard, innerContentList, innerContentListItem } from 'theme/contentEditionCommon/innerContentListTheme';
import ConceptLessonListItem from './ConceptLessonListItem';

const ListCard = styled(StyledListItem)(
	{
		...innerContentListItem,
	}
);

/**
 * @name ConceptLessonList
 * @description A component used in the concept edition page to display its lessons and a "create new lesson" link.
 *
 * @author TimothÃ©e Simon-Franza
 *
 * @param {Array}	lessonList	The lesson list of the concept currently being edited.
 * @param {string}	conceptId	The id of the parent concept.
 */
const ConceptLessonList = ({ lessonList, conceptId }) => {
	const { t } = useTranslation();

	const onAddLessonCardClick = useCallback(() => {
		history.push(generatePath(routes.concepts.lessonCreation, { conceptId }));
	}, [conceptId]);

	return (
		<StyledDiv marginTop="64px">
			<DoublePica as="h3">{t('concepts.edition.lessons_list.title')}</DoublePica>
			<StyledList {...innerContentList}>
				{lessonList.map((lesson) => (
					<ConceptLessonListItem key={lesson.id} {...lesson} />
				))}
				<ListCard {...addInnerContentCard} onClick={onAddLessonCardClick}>
					<PlusCircle />
					{t('concepts.edition.lessons_list.links.add')}
				</ListCard>
			</StyledList>
		</StyledDiv>
	);
};

ConceptLessonList.propTypes = {
	lessonList: PropTypes.arrayOf(
		PropTypes.shape({
			description: PropTypes.string.isRequired,
			id: PropTypes.string.isRequired,
			title: PropTypes.string.isRequired,
		}),
	).isRequired,
	conceptId: PropTypes.string.isRequired,
};

export default ConceptLessonList;
