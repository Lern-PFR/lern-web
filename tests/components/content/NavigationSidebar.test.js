import { shallow } from 'enzyme';
import { NavigationSidebar } from 'components/content';

describe('NavigationSidebar', () => {
	it('should match previous snapshot', () => {
		const sut = (<NavigationSidebar />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});
});
