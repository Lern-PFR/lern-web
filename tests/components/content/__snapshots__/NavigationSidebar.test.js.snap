// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationSidebar should match previous snapshot with items returned from the selector 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <NavigationSidebar />
</ContextProvider>
`;

exports[`NavigationSidebar should match previous snapshot with no item returned from the selector 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <NavigationSidebar />
</ContextProvider>
`;
